lib/std/sort/pdqsort.zig

```diff
@@ -0,0 +1,560 @@
+            } else {
+                block_left = rem / 2;
+                block_right = rem - block_left;
+            }
+            std.debug.assert(block_left <= BLOCK_LEN and block_right <= BLOCK_LEN);
+        }
+
+        if (start_left == end_left) {
+            start_left = 0;
+            end_left = 0;
+
+            var elem: usize = left;
+            var i: usize = 0;
+            while (i < block_left) : (i += 1) {
+                offsets_left[end_left] = left + i;
+
+                if (compareFn(context, items[elem], pivot) != .lt) end_left += 1;
+
+                elem += 1;
+            }
+        }
+
+        if (start_right == end_right) {
+            start_right = 0;
+            end_right = 0;
+
+            var elem: usize = right;
+            var i: usize = 0;
+            while (i < block_right) : (i += 1) {
+                elem -= 1; // avoid owerflow
+
+                offsets_right[end_right] = right - i - 1;
+
+                if (compareFn(context, items[elem], pivot) == .lt) end_right += 1;
+            }
+        }
+
+        const count = math.min(end_left - start_left, end_right - start_right);
+        if (count > 0) {
+            cyclicSwap(
+                T,
+                items,
+                offsets_left[start_left .. start_left + count],
+                offsets_right[start_right .. start_right + count],
+            );
+            start_left += count;
+            start_right += count;
+        }
+
+        if (start_left == end_left) left += block_left;
+
+        if (start_right == end_right) right -= block_right;
+
+        if (is_done) break;
+    }
+
+    if (start_left < end_left) {
+        while (start_left < end_left) {
+            end_left -= 1;
+            mem.swap(T, &items[offsets_left[end_left]], &items[right - 1]);
+            right -= 1;
+        }
+        return right;
+    } else if (start_right < end_right) {
+        while (start_right < end_right) {
+            end_right -= 1;
+            mem.swap(T, &items[left], &items[offsets_right[end_right]]);
+            left += 1;
+        }
+        return left;
+    } else {
+        return left;
+    }
+}
+
+fn cyclicSwap(comptime T: type, items: []T, is: []usize, js: []usize) void {
+    const count = is.len;
+    const tmp = items[is[0]];
+    items[is[0]] = items[js[0]];
+
+    var i: usize = 1;
+    while (i < count) : (i += 1) {
+        items[js[i - 1]] = items[is[i]];
+        items[is[i]] = items[js[i]];
+    }
+
+    items[js[count - 1]] = tmp;
+}
+
+fn breakPatterns(comptime T: type, items: []T) void {
+    @setCold(true);
+    if (items.len < 8) return;
+
+    var r = @intCast(u64, items.len);
+
+    const modulus = math.ceilPowerOfTwoAssert(usize, items.len);
+
+    var idxs: [3]usize = undefined;
+    idxs[0] = items.len / 4 * 2 - 1;
+    idxs[1] = items.len / 4 * 2;
+    idxs[2] = items.len / 4 * 2 + 1;
+
+    for (idxs) |idx| {
+        r ^= r << 13;
+        r ^= r << 17;
+        r ^= r << 5;
```

Comment:
This looks like [xorshift32](https://en.wikipedia.org/wiki/Xorshift#Example_implementation). If so, it should be `r >> 17`

Reply:
yes it's a xorshift. thanks